perfkitbenchmarker/object_storage_service.py

```diff
@@ -0,0 +1,241 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""An interface to object storage services."""
+
+import abc
+import os
+
+from perfkitbenchmarker import errors
+from perfkitbenchmarker import flags
+
+flags.DEFINE_string('object_storage_credential_file', None,
+                    'Directory of credential file.')
+flags.DEFINE_string('boto_file_location', None,
+                    'The location of the boto file.')
+
+FLAGS = flags.FLAGS
+
+DEFAULT_BOTO_LOCATION = '~/.boto'
+BOTO_LIB_VERSION = 'boto_lib_version'
+
+_OBJECT_STORAGE_REGISTRY = {}
+
+
+def GetObjectStorageClass(storage_name):
+  """Return the ObjectStorageService subclass corresponding to 'storage_name'"""
+
+  return _OBJECT_STORAGE_REGISTRY[storage_name]
+
+
+class AutoRegisterObjectStorageMeta(abc.ABCMeta):
+  def __init__(cls, name, bases, dct):
+    super(AutoRegisterObjectStorageMeta, cls).__init__(name, bases, dct)
+    if cls.STORAGE_NAME in _OBJECT_STORAGE_REGISTRY:
+      raise Exception('Duplicate storage implementations for name "%s"' %
+                      cls.STORAGE_NAME)
+    _OBJECT_STORAGE_REGISTRY[cls.STORAGE_NAME] = cls
+
+
+class ObjectStorageService(object):
+  __metaclass__ = AutoRegisterObjectStorageMeta
+
+  STORAGE_NAME = None
+
+  # Keeping the location in the service object is not very clean, but
+  # a nicer solution would be more complex, and we only use different
+  # locations in a very limited way. Also, true multi-location
+  # providers would require another abstraction for Azure service
+  # accounts, which would add more complexity.
+
+  # Service object lifecycle
+
+  def PrepareService(self, location):
+    """Get ready to use object storage.
+
+    This method should be called before any other method of this
+    class. Once it is called, all of this class' methods should work
+    with data in the given location.
+
+    Args:
+      location: where to place our data.
+    """
+    pass
+
+  def CleanupService(self):
+    """Clean up what we did.
+
+    No other method of this class should be called after
+    CleanupProvider.
+    """
+    pass
+
+  # Bucket management
+
+  @abc.abstractmethod
+  def MakeBucket(self, bucket):
+
+    """Make an object storage bucket.
+
+    Args:
+      bucket: the name of the bucket to create.
+    """
+    pass
+
+  @abc.abstractmethod
+  def DeleteBucket(self, bucket):
+    """Delete an object storage bucket.
+
+    Args:
+      bucket: the name of the bucket to delete.
+
+    This method should succeed even if bucket contains objects.
+    """
+    pass
+
+  @abc.abstractmethod
+  def EmptyBucket(self, bucket):
+    """Empty an object storage bucket.
+
+    Args:
+      bucket: the name of the bucket to empty.
+    """
+    pass
+
+  # Working with a VM
+
+  def PrepareVM(self, vm):
```

Comment:
If they are abstract, then any subclass has to implement them. I wanted to make it possible for subclasses to not implement them if they have no preparation or cleanup to do.